### ISSUE TITLE
Copy data from TPU to CPU increases CompileTime

### ISSUE THREAD

**Issue by fengyang0317** [pytorch/xla#7943]
## 🐛 Bug

`.cpu()` increases the CompileTime and makes the code much slower. Not matter the order of `.cpu()` and `xm.mark_step()`.

## To Reproduce

Run the colab https://colab.research.google.com/drive/1Ixqbr2ejpcZpXcx4v_Bt7UEJd71gfpcJ?usp=sharing

Sample code
```
import time

import numpy as np
import torch
import torch_xla
import torch_xla.core.xla_model as xm
import torch_xla.debug.metrics as met
import torch_xla.distributed.parallel_loader as pl
import torch_xla.distributed.spmd as xs
import torch_xla.utils.utils as xu
import torchvision
from torch_xla import runtime as xr

torch_xla.runtime.use_spmd()

loader = xu.SampleGenerator(
  data=torch.zeros(128, 3, 224, 224),
  sample_count=205)

device = xm.xla_device()
num_devices = xr.global_runtime_device_count()
device_ids = np.arange(num_devices)
mesh_shape = (num_devices, 1, 1, 1)
input_mesh = xs.Mesh(device_ids, mesh_shape, ('B', 'C', 'W', 'H'))

loader = pl.MpDeviceLoader(
  loader,
  device,
  input_sharding=xs.ShardingSpec(input_mesh, (0, 1, 2, 3)))

model = torchvision.models.resnet50().to(device)

for i, data in enumerate(loader):
  if i == 5:
    start_time = time.perf_counter()
  output = model(data)
  xm.mark_step()

end_time = time.perf_counter()
elapsed_time = end_time - start_time
print(f"Elapsed time: {elapsed_time:.2f} seconds")

print(met.short_metrics_report())
```

Without `.cpu()`
```
Elapsed time: 7.38 seconds
Counter: CachedCompile
  Value: 203
Metric: CompileTime
  TotalSamples: 2
```

`.cpu()` before `xm.mark_step()`
```
Elapsed time: 18.28 seconds
Counter: CachedCompile
  Value: 407
Metric: CompileTime
  TotalSamples: 208
```
`.cpu()` after `xm.mark_step()`
```
Elapsed time: 16.96 seconds
Counter: CachedCompile
  Value: 203
Metric: CompileTime
  TotalSamples: 207
```

## Expected behavior

Is it possible to avoid the increase of CompileTime?

## Environment

 - Reproducible on XLA backend [CPU/TPU/CUDA]: TPU
 - torch_xla version: nightly

**Comment by JackCaoG** [comment]
@will-cromar do you want to take this one?

**Comment by will-cromar** [comment]
There are two reasons the elapsed time goes way up if you call `.cpu` after `mark_step`:

1. `mark_step` is async -- it marks the point where we "cut" the pending graph and dispatch pending lazy operations to the runtime for execution. It does _not_ block the main thread during device execution. Your baseline without `.cpu()` is queueing up several executables in the background, and the TPU is never waiting for the next iteration. `.cpu()` necessarily blocks the main thread until execution is finished and the data we need to move back to the host is ready. You can block execution without additional side-effects (see below) by putting `xm.wait_device_ops()` after each `mark_step`. When I tried this, the total time went up to ~13 seconds from <10.
2. `cpu()` causes compilation in the background -- when you pass sharded data through a replicated model like this, you end up with a sharded output. That is, there is a separate shard of the output on each device. In practice, these shards may be split over multiple hosts or processes, and we cannot always access each shard directly. `cpu()` on a sharded tensor means we must move the global tensor to the CPU. We run a [short executable](https://github.com/pytorch/xla/blob/989ac69cf4ff913fe4c76329f7447a7319701941/torch_xla/csrc/runtime/pjrt_computation_client.cc#L333-L394) in the background to stage the global tensor data on a local device before we move it to the host. That's where the additional compilation time comes from. If you want to just move the individual local shards to the CPU, you can use `XlaShardedTensor.local_shards`. This is what I ended up with, which also takes ~13 seconds in my testing:

```
  xm.mark_step()
  xm.wait_device_ops()
  # Output is already sharded in the background, but we need to "downcast" it
  output_spmd = torch_xla.distributed.spmd.XLAShardedTensor(output)
  output_spmd.local_shards
```

**Comment by fengyang0317** [comment]
Thanks for your anser. Based on your answer, I tried 
```
  xs.mark_sharding(output, input_mesh, (None, None))
  cpu = output.cpu()
  xm.mark_step()
```

The output is 
```
Elapsed time: 13.46 seconds
Counter: CachedCompile
  Value: 407
Metric: CompileTime
  TotalSamples: 3
```

The CompileTime does not increase anymore, which looks better. While, the running time is still increased from 7 seconds to 13 second. The output is just a 128x1000 tensor. Is it the data transfer that takes extra time?

**Comment by will-cromar** [comment]
I don't think the data transfer itself is taking a significant amount of time -- the difference between (1) and (2) was negligible in my comment above, and the only difference was whether I moved data to the CPU. What really increases the elapsed time is blocking on every iteration.

Tracing (ie running your code and recording operations lazily) and dispatching a new executable is surprisingly slow. The ideal case for performance is overlapping tracing and device execution. That is, you want to start device execution with `xm.mark_step` for the current step, then let the main thread continue along to trace/enqueue the following steps[^1] in parallel with device execution. Doing something to that blocks the main thread (usually by calling `cpu()` IME) can be very costly: the device must now sit idle while the main thread goes and traces the next step and incurs overhead associated with compilation/execution dispatch. In practice, you would need to reduce the blocking operations in your main loop[^2] for best results, e.g. by reducing the frequency of metric updates and checkpoints.

[^1]: The number of async executions we can enqueue is controlled by `XLA_TPU_MAX_INFLIGHT_COMPUTATIONS` on TPU
[^2]: We have a section in our troubleshooting doc about debugging accidental recompilations/executions for less obvious cases: https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#pytorchxla-debugging-tool

You can see the difference in device idle time very concretely with the new `tpu-info` CLI tool on Colab. The duty cycle percentage (basically the amount of time the TPU chip is doing any work) is much higher without blocking than with them:

Baseline:

![image](https://github.com/user-attachments/assets/1edbb003-1019-4cfd-a65b-4dcd7f5a8136)

With `cpu()`:

![image](https://github.com/user-attachments/assets/77d0248c-705c-4f02-95ef-70fae53a7253)

With `wait_device_ops()`:

![image](https://github.com/user-attachments/assets/d8a02aa3-526f-4078-a528-94ce4b9cab8b)

`tpu-info` is available in the terminal sidebar on Colab. On other machines, you can install it with `pip install tpu-info` as of last week.
